org.eclipse.scanning.example/src/org/eclipse/scanning/example/detector/PosDetector.java

```diff
@@ -0,0 +1,161 @@
+import org.eclipse.scanning.api.points.IPosition;
+import org.eclipse.scanning.api.scan.ScanningException;
+import org.eclipse.scanning.api.scan.rank.IScanRankService;
+import org.eclipse.scanning.api.scan.rank.IScanSlice;
+import org.eclipse.scanning.example.Services;
+
+public class PosDetector extends AbstractRunnableDevice<PosDetectorModel> implements IWritableDetector<PosDetectorModel>, INexusDevice<NXdetector> {
+
+	private IDataset image;
+	private ILazyWriteableDataset data;
+	private ILazyWriteableDataset uniqueKeys;
+	private int scanRank = -1;
+	private String filePath;
+	
+	public PosDetector() throws ScanningException {
+		super(Services.getRunnableDeviceService());
+		this.model = new PosDetectorModel();
+		setDeviceState(DeviceState.READY);
+	}
+	
+	@Override
+	public void run(IPosition position)
+			throws ScanningException, InterruptedException, TimeoutException, ExecutionException {
+		image = Random.rand(64, 64);
+	}
+
+	@Override
+	public boolean write(IPosition position) throws ScanningException, InterruptedException {
+		if (image == null) {
+			return false;
+		}
+		
+		if (data == null) {
+			createNexusFile();
+		}
+		
+		try {
+			IScanSlice rslice = IScanRankService.getScanRankService().createScanSlice(position, 64, 64);
+			SliceND sliceND = new SliceND(data.getShape(), data.getMaxShape(),
+					rslice.getStart(), rslice.getStop(), rslice.getStep());
+			data.setSlice(null, image, sliceND);
+			
+			// write unique key
+			final int uniqueKey = position.getStepIndex() + 1;
+			final Dataset uniqueKeyDataset = DatasetFactory.createFromObject(uniqueKey);
+			rslice = IScanRankService.getScanRankService().createScanSlice(position);
+			sliceND = new SliceND(uniqueKeys.getShape(), uniqueKeys.getMaxShape(),
+					rslice.getStart(), rslice.getStop(), rslice.getStep());
+			uniqueKeys.setSlice(null, uniqueKeyDataset, sliceND);
+			
+		} catch (DatasetException e) {
+			setDeviceState(DeviceState.FAULT);
+			throw new ScanningException("Failed to write the data to the NeXus file", e);
+		}
+		
+		setDeviceState(DeviceState.ARMED);
+		return true;
+	}
+	
+	private void createNexusFile() throws ScanningException {
+		TreeFile treeFile = NexusNodeFactory.createTreeFile(filePath);
+		NXroot root = NexusNodeFactory.createNXroot();
+		treeFile.setGroupNode(root);
+		NXentry entry = NexusNodeFactory.createNXentry();
+		root.setEntry(entry);
+		
+		NXcollection ndAttributesCollection = NexusNodeFactory.createNXcollection();
+		entry.setCollection("NDAttributes", ndAttributesCollection);
+		uniqueKeys = ndAttributesCollection.initializeLazyDataset("NDArrayUniqueId", scanRank, Integer.class);
+		
+		NXdata dataGroup = NexusNodeFactory.createNXdata();
+		entry.setData(getName(), dataGroup);
+		data = dataGroup.initializeLazyDataset(NXdata.NX_DATA, scanRank + 2, Double.class);
+		
+		INexusFileFactory nff = ServiceHolder.getNexusFileFactory();
+		NexusFile file = nff.newNexusFile(filePath, true);
```

Comment:
![BLOCKER](https://sonarsource.github.io/sonar-github/severity-blocker.png 'Severity: BLOCKER') Use try-with-resources or close this "NexusFile" in a "finally" clause. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=squid%3AS2095)